torch/csrc/jit/passes/peephole_list_idioms.cpp

```diff
@@ -21,17 +22,312 @@ c10::optional<size_t> normalizeIndex(int64_t index, size_t len) {
+            continue;
+          }
+          if (true_block_throws) {
+            block_refinements.insert(
+                false_block_refinements.begin(), false_block_refinements.end());
+          } else {
+            block_refinements.insert(
+                true_block_refinements.begin(), true_block_refinements.end());
+          }
+          Block* non_throwing_block =
+              true_block_throws ? n->blocks().at(1) : n->blocks().at(0);
+          for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+            if (info_.count(non_throwing_block->outputs().at(i))) {
+              info_[n->outputs().at(i)] =
+                  info_[non_throwing_block->outputs().at(i)];
+            }
+          }
+          continue;
+        }
+
+        // if either block has a constant bool output, e.g. `true` on the
+        // truee block, then for the `false` value we can take the false
+        // refinements from the other block and from the other block value bc
+        // if the output is false it had to have come from the false block.
+        // Otherwise, just take intersection of refinements
+
+        for (size_t i = 0; i < if_n.outputs().size(); ++i) {
+          if (!(if_n.outputs().at(i)->type() == BoolType::get())) {
+            continue;
+          }
+          Value* true_v = if_n.thenOutputs().at(i);
+          Value* false_v = if_n.elseOutputs().at(i);
+
+          if (!info_.count(true_v) && !info_.count(false_v)) {
+            continue;
+          }
+
+          BoolRefinements out;
+          if (auto maybe_bool = constant_as<bool>(true_v)) {
+            if (*maybe_bool) {
+              out = BoolRefinements::FalseRefinements(unionRefinements(
+                  info_[false_v].false_refine(), false_block_refinements));
+            } else {
+              out = BoolRefinements::TrueRefinements(unionRefinements(
+                  info_[false_v].true_refine(), false_block_refinements));
+            }
+          } else if (auto maybe_bool = constant_as<bool>(false_v)) {
+            if (*maybe_bool) {
+              out = BoolRefinements::FalseRefinements(unionRefinements(
+                  info_[true_v].false_refine(), true_block_refinements));
+            } else {
+              out = BoolRefinements::TrueRefinements(unionRefinements(
+                  info_[true_v].true_refine(), true_block_refinements));
+            }
+          }
+          if (info_.count(true_v) && info_.count(false_v)) {
+            out = info_[true_v].intersectBoolRefinements(info_[false_v]);
+          }
+          info_[if_n.outputs().at(i)] = out;
+        }
+      }
+    }
+    active_refinements_.pop_back();
+    return block_refinements;
+  };
+
+  c10::optional<int64_t> tryFindRefinement(Value* v) {
+    for (const auto& ref : active_refinements_) {
+      auto maybe_refinement = ref->find(v);
+      if (maybe_refinement != ref->end()) {
+        return maybe_refinement->second;
+      }
+    }
```

Comment:
Is it possible to have several refinements for the same value? Here we only return the first one.

Reply:
It's not. For that to happen you would need two different contradicting refinements which would be unreachable code
```
if len(x) == 4:
    if len(x) == 5:
          # unreachable code 
```